nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+use http::StatusCode;
+use nexus_db_model::{Sled, SledInstance};
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::pagination::Paginator;
+use nexus_db_queries::db::DataStore;
+use nexus_types::identity::Asset;
+use oximeter::types::ProducerRegistry;
+use sled_agent_client::Client as SledAgentClient;
+use std::fmt;
+use std::future::Future;
+use std::net::IpAddr;
+use std::num::NonZeroU32;
+use std::sync::Arc;
+use std::sync::Mutex;
+use uuid::Uuid;
+
+/// Background task that periodically checks instance states.
+#[derive(Clone)]
+pub(crate) struct InstanceWatcher {
+    datastore: Arc<DataStore>,
+    resolver: internal_dns::resolver::Resolver,
+    metrics: Arc<Mutex<metrics::Metrics>>,
+}
+
+const MAX_SLED_AGENTS: NonZeroU32 = unsafe {
+    // Safety: last time I checked, 100 was greater than zero.
+    NonZeroU32::new_unchecked(100)
+};
+
+impl InstanceWatcher {
+    pub(crate) fn new(
+        datastore: Arc<DataStore>,
+        resolver: internal_dns::resolver::Resolver,
+        producer_registry: &ProducerRegistry,
+    ) -> Self {
+        let metrics = Arc::new(Mutex::new(metrics::Metrics::default()));
+        producer_registry
+            .register_producer(metrics::Producer(metrics.clone()))
+            .unwrap();
+        Self { datastore, resolver, metrics }
+    }
+
+    fn check_instance(
+        &self,
+        opctx: &OpContext,
+        sled: &Sled,
+        client: &SledAgentClient,
+        instance: SledInstance,
+    ) -> impl Future<Output = CheckResult> + Send + 'static {
+        let instance_id = instance.instance_id();
+        let watcher = self.clone();
+        let target = InstanceTarget {
+            instance_id,
+            sled_agent_id: sled.id(),
+            sled_agent_ip: std::net::Ipv6Addr::from(sled.ip).into(),
+            sled_agent_port: sled.port.into(),
+        };
+        let opctx = opctx.child(
+            std::iter::once((
+                "instance_id".to_string(),
+                instance_id.to_string(),
+            ))
+            .collect(),
+        );
+        let client = client.clone();
+
+        async move {
+            let InstanceWatcher { datastore, resolver, .. } = watcher;
+            slog::trace!(opctx.log, "checking on instance...");
+            let rsp = client.instance_get_state(&instance_id).await;
+            let state = match rsp {
+                Ok(rsp) => rsp.into_inner(),
+                Err(ClientError::ErrorResponse(rsp)) => {
+                    let status = rsp.status();
+                    if status == StatusCode::NOT_FOUND
+                        && rsp.as_ref().error_code.as_deref()
+                            == Some("NO_SUCH_INSTANCE")
+                    {
+                        slog::info!(opctx.log, "instance is wayyyyy gone");
+                        // TODO(eliza): eventually, we should attempt to put the
+                        // instance in the `Failed` state here.
```

Comment:
I was a bit uncertain about whether it was safe to do that here. I was a bit worried about potentially racing with other Nexus replicas that may have received a different result from checking on the same instance and sled-agent, which I thought Greg's "instance updater lock" would resolve; I thought it was better to wait until I'd implemented that. For the other case where we update instance states, we've _received_ a `Generation` from the sled-agent, so that prevents racy updates, but here, we would have to increment it ourselves.

If you think it would be safe to do it now, though, I could be convinced to go ahead and add it here!

Reply:
I think that is a fear which is warranted -- we'll need to fix it eventually, but we don't *need* to fix it here. Let's punt to a different PR.

Reply:
Great, let's do that, then.